### ISSUE TITLE
QoE Metrics Reporting - Add configuration to Service Access Information at API reference point M5

### ISSUE THREAD

**Issue by dsilhavy** [5G-MAG/rt-5gms-application-function#95]
# Introduction

Now that the configuration of QoE Metrics Reporting has been implemented for the Provisioning Session (#87 ), the configuration needs exposing as part of the ServiceAccessInformation on the interface at reference point M5.

This will instruct the Media Session Handler on the UE to collect QoE metrics and submit them to the 5GMS Application Function using the Qoe Metrics Reporting API on the interface at reference point M5 (see #76).

# Design

The `ServiceAccessInformation.clientMetricsReportingConfigurations` field will be added to the `ServiceAccessInformation` response at M5 if a `MetricsReportingConfiguration` is currently set in the provisioning session using the *MetricsReportingProvisioning* API interface at reference point M1. The field will be omitted if there is currently no `MetricsReportingConfiguration` configured in the provisioning session.

# References

TS 26.512 (v17.6.0)
- Clauses 4.7.2 & 11.2
- Annex C.4.1

**Comment by stojkovicv** [comment]
Here is an overview of the metrics related values for the Provisioning API at reference point M1 and Service Access Information at M5.

| MetricsReportingConfiguration | clientMetricsReportingConfiguration | Notes |
| --- | --- | --- |
|`metricsReporting‌ConfigurationId` | `metricsReporting‌ConfigurationId` | This value within clientMeticsReportingConfiguration is the same as corresponding identifier at reference point M1.
|`n/a` | `serverAddresses` | «A list of 5GMS AF addresses to which metrics reports shall be sent. Each address shall be an opaque base URL, following the 5GMS URL format specified in clause 6.1 up to and including the {apiVersion} path element.»|
|`scheme` | `scheme` | Same for metricsReportingConfiguration as well as for clientMetricsReportingConfiguration. If not present - the 3GPP metrics scheme urn:3GPP:ns:PSS:DASH:QM10 from TS 26.247 shall apply.|
|`dataNetworkName` | `dataNetworkName` | Same values, provided by M1 Interface. If not specified – the default DNN shall be used.|
|`reportingInterval` | `reportingInterval` | Value provided by M1 MetricsReportingConfiguration. If not specified, a single final report shall be sent after the media streaming session has ended.|
|`samplePercentage` | `samplePercentage` | Value provided by M1 MetricsReportingConfiguration, between 0.0 and 100.0. If not specified, reports shall be sent for all Sessions.|
|`urlFilters` | `urlFilters` | List of URL patterns for which metrics reporting shall be done. |
|`metrics` | `metrics` | TBD|

I added the code for the metrics related endpoint and http `POST` requests at M5 interface. Current version is parsing JSON body data - however after internal clarification with @dsilhavy, I understood that the streaming metrics are available in .xml format, so I have to work on adapting my implementation. I will also expand current code meant to write all collected metrics to the disk as agreed in #90
Also, pure metrics will be noted further, after my investigation on the client side.

**Comment by stojkovicv** [comment]
Note that this description are provided according to the version TS 26.512 V17.5.0 as I, apparently, don't have access to the latest 17.6.0.

**Comment by dsilhavy** [comment]
> Note that this description is provided according to the version TS 26.512 V17.5.0 as I, apparently, don't have access to the latest 17.6.0.

17.6.0 should be available here: https://portal.3gpp.org/desktopmodules/Specifications/SpecificationDetails.aspx?specificationId=3647 , check versions tab

**Comment by stojkovicv** [comment]
Right now it looks like this:
```
{
    "provisioningSessionId": "0f3b7dae-6115-41ee-8d03-33801f61ce92",
    "provisioningSessionType": "DOWNLINK",
    "streamingAccess": {
        "entryPoints": [
            {
                "locator": "http://localhost/m4d/provisioning-session-0f3b7dae-6115-41ee-8d03-33801f61ce92/BigBuckBunny_4s_onDemand_2014_05_09.mpd",
                "contentType": "application/dash+xml"
            }
        ]
    },
    "clientMetricsReportingConfigurations": [
        {
            "metricsReportingConfigurationId": "1fa3a662-6115-41ee-8d03-33801f61ce92",
            "scheme": "scheme1",
            "dataNetworkName": "my_network1",
            "reportingInterval": 60,
            "samplePercentage": 0.5,
            "urlFilters": [
                "filter1",
                "filter11",
                "filter111"
            ],
            "metrics": [
                "metric1",
                "metric11",
                "metric111"
            ]
        }
    ]
}
```
It will contain `samplingPeriod` as soon as M1 code is updated.

**Comment by dsilhavy** [comment]
Thanks, I think the only thing missing is `serverAddresses`. This is a mandatory property, see 26.512 11.2.3.1

**Comment by stojkovicv** [comment]
Correct, thanks for pointing out. Data model for `MetricsReportingConfiguration` at M1 Provisioning API doesn't include `serverAddresses`. I might be wrong but I believe it was omitted during TS updates.
If the SAI is populated with the data created at M1, how do we get `serverAdresses` since it's not part of its data model?

**Comment by dsilhavy** [comment]
> Correct, thanks for pointing out. Data model for `MetricsReportingConfiguration` at M1 Provisioning API doesn't include `serverAddresses`. I might be wrong but I believe it was omitted during TS updates. If the SAI is populated with the data created at M1, how do we get `serverAdresses` since it's not part of its data model?

Thats a good question. @davidjwbbc has implemented the same for consumption reporting in #94 to address #93. @davidjwbbc can you point @stojkovicv to the logic that adds `serverAddresses` to the ServiceAccessInformation?

**Comment by davidjwbbc** [comment]
The code that creates the serverAddresses list for the consumption reporting is at:
https://github.com/5G-MAG/rt-5gms-application-function/blob/development/src/5gmsaf/service-access-information.c#L66-L73

**Comment by stojkovicv** [comment]
This is preliminary version of "complete" `clientMetricsReportingConfiguirations` of `ServiceAccessInformation`:
```
{
    "provisioningSessionId": "30dc6b78-62b2-41ee-95a3-670d6ce7885e",
    "provisioningSessionType": "DOWNLINK",
    "streamingAccess": {
        "entryPoints": [
            {
                "locator": "http://localhost/m4d/provisioning-session-30dc6b78-62b2-41ee-95a3-670d6ce7885e/BigBuckBunny_4s_onDemand_2014_05_09.mpd",
                "contentType": "application/dash+xml"
            }
        ]
    },
    "clientMetricsReportingConfigurations": [
        {
            "metricsReportingConfigurationId": "39b73516-62b2-41ee-95a3-670d6ce7885e",
            "scheme": "scheme1",
            "dataNetworkName": "my_network1",
            "reportingInterval": 60,
            "samplePercentage": 0.5,
            "urlFilters": [
                "filter1",
                "filter11",
                "filter111"
            ],
            "samplingPeriod": 2,
            "metrics": [
                "metric1",
                "metric11",
                "metric111"
            ],
            "serverAddresses": [
                "http://(null)/3gpp-m5/v2/"
            ]
        }
    ]
}
```
Code is added to include `serverAddresses`. The value is NULL because of `svr_hostname` being NULL. In order to fix that, I have to provide changes on code within M5 state machine (`msaf-m5-sm.c`) which I completely refactored.

**Comment by stojkovicv** [comment]
| M1 Metrics Provisioning               | TS26.512 v17.7 clientMetricsReportingConfigurations   | M5 OpenAPI SAI Metrics resource         | Types          |
|---------------------------------------|-------------------------------------------------------|--------------------------------------|----------------|
| `metrics_reporting_configuration_id`  | `metricsReportingConfigurationId`                     | `metrics_reporting_configuration_id` | char           |
| *null*                                | `serverAddresses`                                     | `server_addresses`                   | OpenAPI_list_t |
| *null*                                | *null*                                                | `slice_scope`                        | OpenAPI_list_t |
| `scheme`                              | `scheme`                                              | `scheme`                             | char           |
| `data_network_name`                   | `dataNetworkName`                                     | `data_network_name`                  | char           |
| `is_reporting_interval`               | *null*                                                | `is_reporting_interval`              | bool           |
| `reporting_interval`                  | `reportingInterval`                                   | `reporting_interval`                 | int            |
| `is_sample_percentage`                | *null*                                                | *null*                               | bool           |
| `sample_percentage`                   | `samplePercentage`                                    | `sample_percentage`                  | double         |
| `url_filters`                         | `urlFilters`                                          | `url_filters`                        | OpenAPI_list_t |
| `sampling_period`                     | `samplingPeriod`                                      | `sampling_period`                    | int            |
| `metrics`                             | `metrics`                                             | `metrics`                            | OpenAPI_list_t |

**Comment by rjb1000** [comment]
@stojkovicv: I was surprised to see underscore characters in the last column of your table (M5 OpenAPI YAML). Are you sure about that? Looking at table 11.2.3.1-1 of TS 26.512, the _ServiceAccessInformation_ resource at M5 uses the same camelcase naming conventions as those used to provision QoE metrics reporting at reference point M1.

**Comment by rjb1000** [comment]
P.S. for _sliceScope_, there is a Change Request going for approval at SA#103 Plenary to add this missing property to the M1 provisioning API:
- TS 26.512
  - Rel-17: **CR0067r1** "Add Slice Scope for metrics configuration and reporting in stage 3" in [S4-240410](https://www.3gpp.org/ftp/TSG_SA/WG4_CODEC/TSGS4_127_Sophia-Antipolis/Docs/S4-240410.zip).

**Comment by stojkovicv** [comment]
@rjb1000 We are talking about different things, I was, with a third columns, refering to the OpenAPI declarations (Mustache Template) of the Application Function's source code. Predefined structures and functions, pulled with bash generator - meant to be utilized by business logic of Application Function. I am sorry if there are any ambiguities.

The table is supposed to help to spot if there are any differences or discrepancies between TS Data model and OpenAPI declarations.

**Comment by rjb1000** [comment]
Ah! No worries then.

**Comment by dsilhavy** [comment]
Required changes were merged in #143
